polyply/src/map_to_molecule.py

```diff
@@ -151,16 +152,34 @@ def match_nodes_to_blocks(self, meta_molecule):
             self.node_to_block[node] = meta_molecule.nodes[node]["resname"]
 
         # fragment nodes match parts of blocks, which describe molecules
-        # with more than one residue
+        # with more than one residue. Sometimes multiple molecules come
+        # after each other in that case the connected component needs to
+        # be an integer multiple of the block
         for fragment in nx.connected_components(restart_graph):
-            block_name = restart_attr[list(fragment)[0]]
-            if all([restart_attr[node] == block_name  for node in fragment]):
-                self.fragments.append(fragment)
-                for node in fragment:
-                    self.node_to_block[node] = block_name
-                    self.node_to_fragment[node] = len(self.fragments) - 1
+            frag_nodes = list(fragment)
+            block = self.force_field.blocks[restart_attr[frag_nodes[0]]]
+            block_res = make_residue_graph(block, attrs=('resid', 'resname'))
+            len_block = len(block_res)
+            len_frag = len(fragment)
+            # here we check if the fragment is an integer multiple
+            # of the multiresidue block
+            if len_frag%len_block == 0:
+                n_blocks = len_frag//len_block
             else:
-                raise IOError
+                # if it is not raise an error
+                molname = restart_attr[frag_nodes[0]]
+                msg = (f"When mapping the molecule {molname} onto the residue graph "
+                        "nodes labeled with from_itp, a mismatch in the length between "
+                        "the provided molecule and the residue graph is found. Make "
+                        "sure that all residues are in the residue graph and input itp-file.")
+                raise IOError(msg)
+
+            for fdx in range(0, n_blocks):
+                current_frag_nodes = frag_nodes[fdx*len_block: (fdx+1)*len_block]
+                self.fragments.append(current_frag_nodes)
+                for node in current_frag_nodes:
+                    self.node_to_block[node] = restart_attr[frag_nodes[0]]
+                    self.node_to_fragment[node] = len(self.fragments) - 1
```

Comment:
Not sure if this for-loop can become rather long, if so it might be worth creating a  `molname` and `num_fragments` variable and not determining it every time.

Reply:
which for loop do you mean? Starting at 158, 177 or 180?

Reply:
The one starting at  180. I meant for example storing `len(self.fragments) - 1` as a variable `num_fragments`, since you use that value in every iteration of the for-loop.